### PR TITLE
New Lidar Mount and Open3d SLAM 

### PR DESCRIPTION
**Pull request by manthan99**
The goal of this PR is to add relevant launch files and config files for running open3d SLAM in the following 3 modes:

Localization Mode (By using an exisiting pcd file)
Online SLAM (SLAM on the fly)
Build Map from Rosbag
Running instructions for all 3 modes have been added here

https://github.com/ETHZ-RobotX/smb_common/tree/cleanup/open3dslam_loc/smb_slam

We make use of the following branch of Open3dSLAM
https://github.com/leggedrobotics/open3d_slam/tree/robotics_summer_school_2023

Additiionally this PR also modifies the TFs and the URDF files for the new LiDAR mount

TODO:

- [ ] Change the branch name of open3d_slam in the smb.repos 
- [ ] Commenting the smb slam launch files
- [ ] Calibration files for Camera-Lidar (Object Detection)

### CODE REVIEW

**Comment by manthan99** [comment]
@mantelt the changes to the URDF have not been done according to the new topic names, etc so we will have to merge those changes as well

File: smb_lowlevel_controller/smb_lowlevel_controller/config/hardware.yaml
Comment (manthan99) [inline smb_lowlevel_controller/smb_lowlevel_controller/config/hardware.yaml:1]: Not very sure where this change came

File: smb_description/urdf/new_payload/meshes/LidarBoxV2.stl
Comment (mantelt) [inline smb_description/urdf/new_payload/meshes/LidarBoxV2.stl:1]: Do we need the stl file as well? .dae should be ok, isn't it?

File: smb_description/urdf/sensor_plugins/smb_robosense.gazebo.xacro [hunk 1]
```diff
@@ -13,7 +13,7 @@
       <xacro:include filename="$(find robosense_description)/urdf/RS-16.urdf.xacro"/>
 
       <xacro:RS-16 parent="${base_link}" name="rslidar" topic="$(arg pointcloud_topic)" hz="10" samples="440" gpu="${gpu}">
-        <origin xyz="0.00445 0 ${0.298+chassisHeight/2.0}" rpy="0 0 ${M_PI/2.0}" />
+        <origin xyz="0.0245 0 ${0.228+0.1+chassisHeight/2.0}" rpy="0 0 0" />
```
Comment (mantelt) [inline smb_description/urdf/sensor_plugins/smb_robosense.gazebo.xacro:16]: would be nice to have parameters for the 0.228 (payload height) and 0.1 (lidar box height).

File: smb_lowlevel_controller/smb_lowlevel_controller/config/hardware.yaml
Comment (mantelt) [inline smb_lowlevel_controller/smb_lowlevel_controller/config/hardware.yaml:1]: as this has nothing to do with the open3d_slam (or does it? I don't see the connection), I would suggest to revert this change.

File: smb_slam/README.md
Comment (mantelt) [inline smb_slam/README.md:1]: thanks for adding this documentation.

File: smb_slam/config/open3d_slam/param_rs16_localization.lua [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+params.mapper_localizer.is_build_dense_map = false
+params.mapper_localizer.is_use_map_initialization = true
+params.mapper_localizer.is_print_timing_information = true
+params.mapper_localizer.scan_to_map_registration.scan_processing.voxel_size = 0.08
+params.mapper_localizer.scan_to_map_registration.scan_processing.downsampling_ratio = 0.25
+params.mapper_localizer.scan_to_map_registration.scan_processing.scan_cropping.cropping_radius_max = 40.0
+params.mapper_localizer.scan_to_map_registration.icp.max_correspondence_dist = 0.8
+
+--MAP_INITIALIZER
+params.map_initializer.pcd_file_path = "/home/robotx/catkin_workspaces/smb_dev/src/core/smb_slam/data/maps/map.pcd"
```
Comment (mantelt) [inline smb_slam/config/open3d_slam/param_rs16_localization.lua:24]: Is it necessary to have a full path here? can we just omit this param and only use a ros param for it? or have it somehow relative to the open3d_slam directory?

File: smb_slam/launch/build_map_from_bag.launch [hunk 3]
```diff
@@ -1,25 +1,39 @@
...
-  <arg name="parameter_folder_path" default="$(find smb_slam)/config/open3d_slam/"/>
-  <arg name="parameter_filename"    default="params_robosense_rs16.yaml"/>
-  <arg name="read_from_rosbag"      value="true"/>
-  <arg name="rosbag_filepath"       default="$(env HOME)/"/> # Can be used to process rosbag sequentially, usually much faster without missing messages
-  <arg name="launch_rviz"           default="false"/>
-  <arg name="use_sim_time"          default="false"/>
+	<arg name="launch_prefix" default="" doc="gdb -ex run --args"/>
+	<arg name="launch_rviz" default="true" />
+	<arg name="cloud_topic" default="/rslidar_points" />
+  <arg name="parameter_filename" default="param_robosense_rs16.lua"/>
```
Comment (mantelt) [inline smb_slam/launch/build_map_from_bag.launch:8]: indentation

File: smb_slam/package.xml [hunk 4]
```diff
@@ -11,10 +11,11 @@
...
-	<depend>pointmatcher_ros</depend>
-	<depend>libpointmatcher</depend>
+	<depend>open3d_slam_ros</depend>
+	<!-- <depend>pointmatcher_ros</depend> -->
+	<!-- <depend>libpointmatcher</depend> -->
 	<depend>roslib</depend>
-	<depend>icp_localization</depend>
+	<!-- <depend>icp_localization</depend> -->
 	
 	<build_depend>libgoogle-glog-dev</build_depend>
```
Comment (mantelt) [inline smb_slam/package.xml:20]: isn't this dep obsolete?

File: smb_slam/launch/localization_open3dslam.launch [hunk 5]
```diff
@@ -2,14 +2,14 @@
...
+  <!-- Without setting the map path, it will throw an error-->
   <arg name="cloud_topic"           default="/rslidar_points"/>
+  <arg name="parameter_filename"    default="param_rs16_localization.lua"/>
   <arg name="parameter_folder_path" default="$(find smb_slam)/config/open3d_slam/"/>
-  <arg name="parameter_filename"    default="params_robosense_rs16.yaml"/>
-  <arg name="read_from_rosbag"      default="false"/>
-  <arg name="rosbag_filepath"       default=""/> # Can be used to process rosbag sequentially, usually much faster without missing messages
   <arg name="launch_rviz"           default="false"/>
   <arg name="launch_sensors"        default="false"/>
-  <arg name="use_sim_time"          default="false"/>
```
Comment (mantelt) [inline smb_slam/launch/localization_open3dslam.launch:12]: one could potentially run the localization in sim. Don't we need this argument then?

**Review by mantelt — CHANGES_REQUESTED** [review]
Hi @manthan99
thanks for setting up this PR. Would be great if you could address my suggestions!

File: smb_slam/launch/build_map_from_bag.launch
Comment (mantelt) [inline smb_slam/launch/build_map_from_bag.launch:1]: Please add documentation for the arguments

File: smb_slam/launch/online_slam.launch
Comment (mantelt) [inline smb_slam/launch/online_slam.launch:1]: Would be great to have [documentation of the arguments](http://wiki.ros.org/roslaunch/XML/arg#Attributes) in the launch file

File: smb_slam/launch/open3d_slam/mapping.launch
Comment (mantelt) [inline smb_slam/launch/open3d_slam/mapping.launch:1]: Please add a doc attribute to all the arguments

File: smb_slam/package.xml (hunk 4 above)
Comment (manthan99) [reply smb_slam/package.xml]: Yes

File: smb_slam/launch/localization_open3dslam.launch (hunk 5 above)
Comment (manthan99) [reply smb_slam/launch/localization_open3dslam.launch:12]: The sim time argument is only used if the rosbag file is run through the launch file. 
For the localization mode, we don't provide an option to run the rosbag file from the same launch file (Since this will be tested live). Alternatively one can still run the rosbag file for localization from a separate terminal so I don't think this is necessary.

File: smb_slam/launch/online_slam.launch
Comment (mantelt) [reply smb_slam/launch/online_slam.launch:1]: In addition to the documentation of the arguments that are still missing:

Wouldn't it make sense to put the online slam and the localization only in one launch file, and switch depending on the argument (i.e. if a pcd file has been provided).

File: smb_slam/README.md [hunk 6]
```diff
@@ -3,7 +3,13 @@ This package contains a bunch of commodity scripts for running (simultaneous) lo
 
 For creating the map, [open3d_slam](https://github.com/leggedrobotics/open3d_slam) can be used. We have created a new branch for the robotics summer school. Make sure that you are using the correct branch ([robotics_summer_school_2023](https://github.com/leggedrobotics/open3d_slam/tree/robotics_summer_school_2023)).
 
-The map generation is both possible in real-time on the system or from a rosbag, as well as by directly reading from the rosbag.
+It is necessary to follow the instructions [here](https://open3d-slam.readthedocs.io/en/latest/installation.html) to install all dependencies before compile `smb_slam`.
```
Comment (mantelt) [inline smb_slam/README.md:6]: >  before compile `smb_slam`.

--> before compiling `smb_slam`

**Review by mantelt — CHANGES_REQUESTED** [review]
thanks for addressing my comments! 

What do you think about merging the online slam and the localization only launch files? Does that make sense?

Otherwise, I think we should soon be good to merge...

File: smb_slam/launch/localization_open3dslam.launch
Comment (mantelt) [inline smb_slam/launch/localization_open3dslam.launch:1]: when launching the localization only using this launch file like this:
```(bash)
roslaunch smb_slam localization_open3dslam.launch pcd_file_path:=/home/thomas/catkin_workspaces/smb_rss_dev/src/core/smb_slam/maps/gazebo_wangen.pcd
```

I get the following error:
```
RLException: unused args [pcd_file_path] for include of [/home/thomas/catkin_workspaces/smb_open3d_slam/src/slam/open3d_slam/open3d_slam_ros/launch/mapping.launch]
The traceback for the exception was written to the log file
```

Comment (manthan99) [reply smb_slam/launch/localization_open3dslam.launch:1]: Yes you will need to pull changes fo the robotics summer school branch for open3d slam

Comment (manthan99) [reply smb_slam/launch/localization_open3dslam.launch:1]: https://github.com/leggedrobotics/open3d_slam/tree/robotics_summer_school_2023

File: smb_slam/launch/online_slam.launch
Comment (manthan99) [reply smb_slam/launch/online_slam.launch:1]: I would prefer to keep it different for simplicity (even though it means having duplicate files)

**Comment by manthan99** [comment]
@mantelt seems to be done from my side. You can have a look and do the merging whenever possible.
